File: Python/LC106_Const_Binary_Tree_In_Post_Ord.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import numpy as np


class Solution:
    # Runtime: 368 ms, faster than 9.53% of Python3 online submissions for Construct Binary Tree from Inorder and Postorder Traversal.
    # Memory Usage: 32.6 MB, less than 88.89% of Python3 online submissions for Construct Binary Tree from Inorder and Postorder Traversal.
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def helper(inorder, si, ei, postorder, sp, ep):
            if si > ei:
                return None
            root = TreeNode(postorder[ep])
            pos_i = np.where(np.array(inorder[si:(ei+1)]) == root.val)[0][0]
            n_l, n_r = pos_i, ei-si-pos_i
            root.left = helper(inorder, si, si+n_l-1, postorder, sp, sp+n_l-1)
            root.right = helper(inorder, si+pos_i+1, si +
                                pos_i+n_r, postorder, sp+pos_i, sp+pos_i+n_r-1)
            return root
        n = len(inorder)
        return helper(inorder, 0, n-1, postorder, 0, n-1)

File: Cpp/LC106_Const_Binary_Tree_In_Post_Ord.cpp
#include <vector>
#include <stack>

using namespace std;

class Solution
{
public:
    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder)
    {
        if (!inorder.size())
            return nullptr;
        int n = inorder.size();
        stack<TreeNode *> roots;
        int ii = n - 1, pi = n - 1;
        TreeNode *root = new TreeNode(postorder[pi--]), *cur = root;
        roots.push(cur);
        while (ii >= 0)
        { //pi can be negative but ii have reach 0
            while (!roots.empty() && inorder[ii] != roots.top()->val)
            {
                // Process roots.top() right subtree
                cur->right = new TreeNode(postorder[pi--]);
                cur = cur->right;
                roots.push(cur);
            }
            cur = roots.top();
            roots.pop(); // Done with cur node right subtree
            if (pi >= 0 && (roots.empty() || inorder[ii - 1] != roots.top()->val))
            { //roots can be empty.
                // cur node has left subtree
                cur->left = new TreeNode(postorder[pi--]);
                cur = cur->left;
                roots.push(cur);
            }
            --ii;
        }
        return root;
    }

    // TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder)
    // {
    //     if (!inorder.size())
    //         return nullptr;
    //     int n = inorder.size();
    //     TreeNode *root = new TreeNode(postorder[n - 1]);
    //     int ii = n - 1, pi = n - 1;
    //     dfs(inorder, postorder, nullptr, root, ii, pi);
    //     return root;
    // }

    // void dfs(vector<int> &inorder, vector<int> &postorder, TreeNode *pre_inorder_root, TreeNode *root, int &ii, int &pi)
    // {
    //     if (ii < 0)
    //         return;
    //     if (inorder[ii] != root->val)
    //     { //right subtree
    //         root->right = new TreeNode(postorder[--pi]);
    //         dfs(inorder, postorder, root, root->right, ii, pi); //notice the change of ii will be passed back
    //     }
    //     if (inorder[ii] == root->val)
    //     {         //left subtree
    //         --ii; //whenever found inorder[ii]==root->val, done with its right subtree, i minus 1
    //         if (pi > 0 && (pre_inorder_root == nullptr || (ii >= 0 && inorder[ii] != pre_inorder_root->val)))
    //         {
    //             root->left = new TreeNode(postorder[--pi]);
    //             dfs(inorder, postorder, pre_inorder_root, root->left, ii, pi); //keep the predecessor of root in inorder traversal.
    //         }
    //     }
    // }
};
